File: src/Javascript/array/59.螺旋矩阵-ii.js
/*
 * @lc app=leetcode.cn id=59 lang=javascript
 *
 * [59] 螺旋矩阵 II
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    let l = 0, r = n - 1, t = 0, b = n - 1, num = 1;
    const count = n * n;
    const mat = Array.from({length: n}, () => Array(n));
    while (num <= count) {
        for (let i = l; i <= r; i++) mat[t][i] = num++;
        t++;
        for (let i = t; i <= b; i++) mat[i][r] = num++;
        r--;
        for (let i = r; i >= l; i--) mat[b][i] = num++;
        b--;
        for (let i = b; i >= t; i--) mat[i][l] = num++;
        l++;
    }
    return mat;
};
// @lc code=end

File: src/Java/practice/first-time/59.螺旋矩阵-ii.java
/*
 * @lc app=leetcode.cn id=59 lang=java
 *
 * [59] 螺旋矩阵 II
 */

// @lc code=start
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int count = n * n, top = 0, bottom = n - 1, left = 0, right = n - 1, num = 1;
        while (count >= num) {
            for (int i = left; i <= right && count >= num; i++) res[top][i] = num++;
            top++;
            for (int i = top; i <= bottom && count >= num; i++) res[i][right] = num++;
            right--;
            for (int i = right; i >= left && count >= num; i++) res[bottom][i] = num++;
            bottom--;
            for (int i = bottom; i >= top && count >= num; i++) res[i][left] = num++;
            left++;
        }
        return res;
    }
}
// @lc code=end

